pymc3/gp/cov.py

```diff
@@ -97,29 +102,69 @@ def __init__(self, factor_list):
             else:
                 self.factor_list.append(factor)
 
-    def merge_factors(self, X, Z=None, diag=False):
+    def merge_factors(self, X, Xs=None, diag=False):
+        # this function makes sure diag=True is handled properly
         factor_list = []
         for factor in self.factor_list:
+
+            # if factor is a Covariance
             if isinstance(factor, Covariance):
-                factor_list.append(factor(X, Z, diag))
-            elif hasattr(factor, "ndim"):
-                if diag:
-                    factor_list.append(tt.diag(factor))
-                else:
-                    factor_list.append(factor)
-            else:
-                factor_list.append(factor)
+                factor_list.append(factor(X, Xs, diag))
+                continue
```

Comment:
I personally find `continue` a bit harder to read than `elif`, others may disagree.